packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/typedarray/ConcreteTypedArrays.kt

```diff
@@ -0,0 +1,154 @@
+package expo.modules.kotlin.typedarray
+
+import expo.modules.kotlin.jni.JavaScriptTypedArray
+
+@Suppress("NOTHING_TO_INLINE")
+private inline fun AnyTypedArray.checkIfInRange(index: Int) {
+  if (index < 0 || index >= length) {
+    throw IndexOutOfBoundsException()
+  }
+}
+
+class Int8Array(private val rawArray: JavaScriptTypedArray)
+  : AnyTypedArray by rawArray, GenericTypedArray<Byte> {
+  override operator fun get(index: Int): Byte {
+    checkIfInRange(index)
+    return readByte(index)
+  }
+
+  override operator fun set(index: Int, value: Byte) {
+    checkIfInRange(index)
+    writeByte(index, value)
+  }
+}
+
+class Int16Array(private val rawArray: JavaScriptTypedArray)
+  : AnyTypedArray by rawArray, GenericTypedArray<Short> {
+  override operator fun get(index: Int): Short {
+    checkIfInRange(index)
+    return read2Byte(index * 2)
+  }
+
+  override operator fun set(index: Int, value: Short) {
+    checkIfInRange(index)
+    write2Byte(index * 2, value)
+  }
+}
+
+class Int32Array(private val rawArray: JavaScriptTypedArray)
+  : AnyTypedArray by rawArray, GenericTypedArray<Int> {
+  override operator fun get(index: Int): Int {
+    checkIfInRange(index)
+    return read4Byte(index * 4)
+  }
+
+  override operator fun set(index: Int, value: Int) {
+    checkIfInRange(index)
+    write4Byte(index * 4, value)
+  }
+}
+
+class Uint8Array(private val rawArray: JavaScriptTypedArray)
+  : AnyTypedArray by rawArray, GenericTypedArray<UByte> {
+  override operator fun get(index: Int): UByte {
+    checkIfInRange(index)
+    return readByte(index).toUByte()
+  }
+
+  override operator fun set(index: Int, value: UByte) {
+    checkIfInRange(index)
+    writeByte(index, value.toByte())
+  }
+}
+
+class Uint8ClampedArray(private val rawArray: JavaScriptTypedArray)
+  : AnyTypedArray by rawArray, GenericTypedArray<UByte> {
+  override operator fun get(index: Int): UByte {
+    checkIfInRange(index)
+    return readByte(index).toUByte()
+  }
+
+  override operator fun set(index: Int, value: UByte) {
+    checkIfInRange(index)
+    writeByte(index, value.toByte())
+  }
+}
+
+class Uint16Array(private val rawArray: JavaScriptTypedArray)
+  : AnyTypedArray by rawArray, GenericTypedArray<UShort> {
+  override operator fun get(index: Int): UShort {
+    checkIfInRange(index)
+    return read2Byte(index * 2).toUShort()
```

Comment:
Maybe we should extract this `2` and others as constants. Sth like:
```kt
private const val BYTE_LENGTH = 1
private const val USHORT_LENGTH = 2
private const val FLOAT_LENGTH = 4
```
etc.

It would be a little bit more clear about what are the offsets we're calculating here.

Btw. doesn't Kotlin have anything like `sizeof(T)`?